src/openfermion/ops/_binary_operator.py

```diff
@@ -0,0 +1,494 @@
+# symbolic binary class for decoder definitions (arxiv 1712.07067)
+
+import copy
+import numpy as np
+
+
+class SymbolicBinaryError(Exception):
+    pass
+
+
+class SymbolicBinary(object):
+    """
+        SymbolicBinaries are non-linear binary functions that are expressed analytically, they are
+        components of decoders that are used to rotate terms in qubit basis to fermionic basis. 
+        
+        A SymbolicBinary represents a sum of terms. Each term can consist of two types of binary variables:
+        (n,'W') where n stands for the qubit index, or (1,'1') for constant 1.
+         
+        The SymbolicBinary class overloads operations for manipulation of binary functions obeying binary rules.
+        1 + 1 = 0
+
+        Example:
+            .. code-block:: python
+
+                bin_fun = SymbolicBinary('1 + w1 w2 + w0 w1')
+                # Equivalently
+                bin_fun = SymbolicBinary([((1,'1'),)] + SymbolicBinary([((1,'W'),(2,'W')),((0,'W'),(1,'W'))])
+
+        """
+    actions = ('1', 'W')
+    action_strings = ('', 'W')
+    action_before_index = True
+    different_indices_commute = True
+
+    def __init__(self, term=None):
+        """
+        Initialize the SymbolicBinary based on term
+
+        Args:
+            term: (str, list, tuple) used for initializing a SymbolicBinary
+        """
+
+        self.terms = []
+
+        # long string input
+        if isinstance(term, str) and '+' in term:
+            self._long_string_init(term)
+            return
+
+        # Zero operator: leave the terms list empty
+        if term is None:
+            return
+
+        # Sequence input
+        elif isinstance(term, tuple) or isinstance(term, list):  # list of tuples of tuples
+            self._parse_sequence(term)
+
+        # String input
+        elif isinstance(term, str):
+            self.terms.append(tuple(self._parse_string(term)))
+
+        # Invalid input type
+        else:
+            raise ValueError('term specified incorrectly.')
+
+        self._check_terms()
+
+    def _check_terms(self):
+        """
+        Ensures all terms obey binary rules
+        Updates terms in place.
+
+        """
+        sorted_input = []
+        for item in self.terms:
+            if len(item) == 0:
```

Comment:
I think it would be nicer to write:
`if len(item):
    binary_sum_rule(sorted_input, tuple(sorted(set(item))))`
There is no need for the negative conditional followed by continue.